packages/cosmic-swingset/src/chain-main.js

```diff
@@ -79,6 +82,71 @@ const makeChainStorage = (call, prefix = '', imp = x => x, exp = x => x) => {
   return storage;
 };
 
+/**
+ * Create a queue backed by chain storage.
+ *
+ * The queue uses the following storage layout, prefixed by `prefix`, such as
+ * `actionQueue.`:
+ * - `<prefix>head`: the index of the first entry of the queue.
+ * - `<prefix>tail`: the index *past* the last entry in the queue.
+ * - `<prefix><index>`: the contents of the queue at the given index.
+ *
+ * For the `actionQueue`, the Cosmos side of the queue will push into the queue,
+ * updating `<prefix>tail` and `<prefix><index>`.  The JS side will shift the
+ * queue, updating `<prefix>head` and reading and deleting `<prefix><index>`.
+ *
+ * Parallel access is not supported, only a single outstanding operation at a
+ * time.
+ *
+ * @param {(obj: any) => any} call send a message to the chain's storage API and
+ * receive a reply
+ * @param {string} [prefix] string to prepend to the queue's storage keys
+ */
+const makeChainQueue = (call, prefix = '') => {
+  const storage = makeChainStorage(call, prefix);
```

Comment:
I'm slightly nervous about the pre-existing format-confusion opportunity in `makeChainStorage`. It'd be nice if `prefix` were forbidden to include some specific delimiter (maybe `:`), and the `${prefix}${key}` concatenation were replaced with `${prefix}:${key}`, so there was no way for someone to mess up and confuse `{ prefix: 'abc', key: 'def' }` with `{ prefix: 'ab', key: 'cdef' }`.

Reply:
I will gladly assert that the `prefix` must either be `''` or end in `'.'`.

Format confusion is avoided by convention because `makeChainStorage` and `makeChainQueue` are closely held by the manager of the namespace (the chain host).  If that host creates collisions in the key namespace, they get what they pay for.  What attack are you concerned about?